### ISSUE TITLE
U61Osal.h

### ISSUE THREAD

**Issue by AndresL71** [izot/lon-driver#1]
Hello!

Missing U61Osal.h

**Comment by joshmarshall1995** [comment]
@AndresL71 Did you compile the driver without issue? I am getting a ton of redeclaration errors, looks like some of the header files are conflicting.

**Comment by AndresL71** [comment]
Nope.  Couldn’t. Got the same results. Enviado desde mi iPhoneEl 19-04-2024, a la(s) 4:01 a. m., Jmarts ***@***.***> escribió:﻿
@AndresL71 Did you compile the driver without issue? I am getting a ton of redefinition errors, looks like some of the header files are conflicting.

**Comment by AndresL71** [comment]
How to Compile and Run the IzoT Lon-Driver for Echelon USB-U10 on Proxmox / Modern Linux

The original codebase suffers from structural C bugs (missing include guards), hardcoded device paths, and incompatibility with modern Linux kernels (6.x+). Here is how we resolved these issues using a dedicated Virtual Machine environment.

Architecture & Environment Setup
1. Avoid LXC, Use a Dedicated VM

The u61 driver compiles a native kernel module (u61.ko). Attempting to inject this into a Proxmox LXC container restricts isolation and can taint the host kernel.

    Solution: Create a dedicated Virtual Machine (VM) in Proxmox.

    Hardware Passthrough: Map the Echelon USB-U10 directly to the VM using USB Vendor/Device ID passthrough.

2. Kernel Compatibility Selection

Modern kernels (e.g., Kernel 6.12 on Debian 13) have deprecated or refactored internal TTY and network tracking APIs (such as netif_rx_ni, tty_register_ldisc, and write-allowed dev->dev_addr), which completely breaks compilation.

    Solution: Use Ubuntu 20.04 LTS (Kernel 5.4) for the bridge VM. It maintains native compatibility with the legacy Linux networking APIs used in this driver.

Step-by-Step Implementation & Bug Fixes
Step 1: Install Dependencies

Prepare the clean Ubuntu environment with standard build tools, systemd development headers, and the matching kernel headers:
Bash

sudo apt update && sudo apt upgrade -y
sudo apt install -y build-essential git cmake libusb-1.0-0-dev linux-headers-generic libsystemd-dev strace

Step 2: Clone the Repository
Bash

git clone https://github.com/izot/lon-driver.git
cd lon-driver/u61

Step 3: Fix the Missing Include Guards Bug in U61Defs.h

The compilation originally fails with multiple redeclaration of enumerator and conflicting types errors. This happens because U61Defs.h is included multiple times down the stream without protection.

    Solution: Inject a #pragma once directive at the very top of the header file:

Bash

cp include/U61Defs.h include/U61Defs.h.orig
echo "#pragma once" > include/U61Defs.h
cat include/U61Defs.h.orig >> include/U61Defs.h

Step 4: Compile and Load the Kernel Module

Compile the driver out-of-tree by explicitly invoking the kernel build system:
Bash

make clean
make -C /lib/modules/$(uname -r)/build M=$PWD modules
sudo make -C /lib/modules/$(uname -r)/build M=$PWD modules_install
sudo depmod -a
sudo modprobe u61

Verification: Running lsmod | grep u61 should now show the module loaded successfully.
Step 5: Patch Hardcoded Paths and Bugs in lonifd

Navigating to the daemon directory (cd ../lonifd), we found that lonifd.c contains copy-paste typos and hardcoded values that completely ignore user-passed CLI arguments:

    Line 93: Typo pointing the U61 driver file to the /sys/module/u50/ path.

    Line 1243: Hardcoded device string forcing the program to look only for /dev/ttyACM0 instead of standard FTDI serial paths like /dev/ttyUSB0.

Fix both issues directly using sed:
Bash

cd ../lonifd
# Fix the sysfs module path from u50 to u61
sed -i '93s/module\/u50/module\/u61/' lonifd.c
# Change the hardcoded fallback device from ttyACM0 to ttyUSB0
sed -i 's/ttyACM0/ttyUSB0/g' lonifd.c

Step 6: Compile and Install the Network Daemon

Compile the lone lonifd.c file using gcc, linking it to the newly installed systemd libraries:
Bash

gcc lonifd.c -I../u61/include -o lonifd -lsystemd
sudo cp lonifd /usr/local/sbin/

Running the Interface

With the driver active and the daemon patched, launch the link connection mapping your virtual interface name to your physical USB port:
Bash

sudo lonifd lon0 /dev/ttyUSB0

Successful Outcome

The driver successfully binds to the FTDI converter, initializes the network device link, and extracts the hardware's embedded Echelon MAC address.

Checking network interfaces via ip a reveals the newly spawned network interface (e.g., lon10), completely active and bound to a local LonTalk-IP subnet:
Plaintext

5: lon10: <BROADCAST,UP,LOWER_UP> mtu 1280 ...
    link/[2384] 04:7c:90:f9:0c:00 brd ff:ff:ff:ff:ff:ff
    inet 44.1.110.125/24 brd 44.1.110.255 scope global lon10

The LonWorks bus traffic can now be captured natively in Linux using tools like tcpdump -i lon10 -XX.
